src/overlays/actors/ovl_En_Gakufu/z_en_gakufu.c

```diff
@@ -34,28 +39,253 @@ const ActorInit En_Gakufu_InitVars = {
+ * until the timer runs out
+ */
+void EnGakufu_DisplayOnTimer(EnGakufu* this, GlobalContext* globalCtx) {
+    if (this->actor.home.rot.x > 0) {
+        this->actor.draw = EnGakufu_Draw;
+        this->actor.home.rot.x--;
+    } else {
+        this->actor.draw = NULL;
+        this->actor.home.rot.x = 0;
+        this->actionFunc = EnGakufu_WaitForTimer;
+    }
+}
+
+/**
+ * Special behaviour for the Milk Bar:
+ * this->actor.home.rot.x acts as a timer. Waits for this timer to be set,
+ * then rerolls the notes and start displaying the notes
+ */
+void EnGakufu_WaitForTimer(EnGakufu* this, GlobalContext* globalCtx) {
+    if (this->actor.home.rot.x > 0) {
+        EnGakufu_ProcessNotes(this);
+        this->actionFunc = EnGakufu_DisplayOnTimer;
+    }
+}
+
+void EnGakufu_DoNothing(EnGakufu* this, GlobalContext* globalCtx) {
+}
+
+s32 EnGakufu_IsWithinRange(EnGakufu* this, GlobalContext* globalCtx) {
+    if (this->actor.xzDistToPlayer < 600.0f) {
+        return true;
+    } else {
+        return false;
+    }
+}
+
+/**
+ * Reward the player with three item drops depending on the time of day
+ */
+void EnGakufu_GiveReward(EnGakufu* this, GlobalContext* globalCtx) {
+    s32 index;
+    s32 i;
+
+    play_sound(NA_SE_SY_CORRECT_CHIME);
+
+    // 24 hours / The total time in a day
+    index = gSaveContext.time * (24.0f / 0x10000);
+    for (i = 0; i < 3; i++) {
+        Item_DropCollectible(globalCtx, &sRewardDropsSpawnPos, sRewardDrops[i + sRewardDropsIndex[index]]);
+    }
+
+    this->actionFunc = EnGakufu_DoNothing;
+}
+
+void EnGakufu_PlayRewardCutscene(EnGakufu* this, GlobalContext* globalCtx) {
+    if (this->actor.cutscene == -1) {
+        EnGakufu_GiveReward(this, globalCtx);
+    } else if (ActorCutscene_GetCanPlayNext(this->actor.cutscene)) {
+        ActorCutscene_StartAndSetUnkLinkFields(this->actor.cutscene, &this->actor);
+        EnGakufu_GiveReward(this, globalCtx);
+    } else {
+        ActorCutscene_SetIntentToPlay(this->actor.cutscene);
+    }
+}
+
+void EnGakufu_WaitForSong(EnGakufu* this, GlobalContext* globalCtx) {
+    if (gSaveContext.eventInf[3] & 2) {
+        if (gSaveContext.eventInf[3] & 4) {
+            gSaveContext.eventInf[3] &= (u8)~2;
+            gSaveContext.eventInf[3] &= (u8)~4;
+
+            this->actionFunc = EnGakufu_PlayRewardCutscene;
+            EnGakufu_PlayRewardCutscene(this, globalCtx);
+            this->actor.draw = NULL;
+        } else if (!EnGakufu_IsWithinRange(this, globalCtx)) {
+            gSaveContext.eventInf[3] &= (u8)~2;
+        }
+    } else if (EnGakufu_IsWithinRange(this, globalCtx)) {
+        gSaveContext.eventInf[3] |= 2;
+    }
+}
+
+void EnGakufu_Update(Actor* thisx, GlobalContext* globalCtx) {
+    EnGakufu* this = THIS;
+
+    this->actionFunc(this, globalCtx);
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Gakufu/func_80AFC960.s")
+void EnGakufu_Draw(Actor* thisx, GlobalContext* globalCtx) {
+    s32 pad;
+    s32 i;
+    EnGakufu* this = THIS;
```

Comment:
Do these have to be in precisely this order?

Reply:
The only thing that needs to happen is for `EnGakufu* this = THIS;` to be last. `pad/globalCtx2` and `i` can be swapped

Reply:
I think put the pad under `i`, then.